File: medbrains/crates/medbrains-core/src/emergency.rs
//! Emergency module domain types.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

// ── Enums ────────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "triage_level", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum TriageLevel {
    Immediate,
    Emergent,
    Urgent,
    LessUrgent,
    NonUrgent,
    Expectant,
    Unassigned,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "er_visit_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum ErVisitStatus {
    Registered,
    Triaged,
    InTreatment,
    Observation,
    Admitted,
    Discharged,
    Transferred,
    Lama,
    Deceased,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "mlc_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum MlcStatus {
    Registered,
    UnderInvestigation,
    OpinionGiven,
    CourtPending,
    Closed,
}

#[derive(Debug, Clone, Serialize, Deserialize, sqlx::Type)]
#[sqlx(type_name = "mass_casualty_status", rename_all = "snake_case")]
#[serde(rename_all = "snake_case")]
pub enum MassCasualtyStatus {
    Activated,
    Ongoing,
    ScalingDown,
    Deactivated,
}

// ── Structs ──────────────────────────────────────────────

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErVisit {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub patient_id: Uuid,
    pub encounter_id: Option<Uuid>,
    pub visit_number: String,
    pub status: ErVisitStatus,
    pub arrival_mode: Option<String>,
    pub arrival_time: DateTime<Utc>,
    pub chief_complaint: Option<String>,
    pub is_mlc: bool,
    pub is_brought_dead: bool,
    pub triage_level: Option<TriageLevel>,
    pub attending_doctor_id: Option<Uuid>,
    pub bay_number: Option<String>,
    pub disposition: Option<String>,
    pub disposition_time: Option<DateTime<Utc>>,
    pub disposition_notes: Option<String>,
    pub admitted_to: Option<String>,
    pub admission_id: Option<Uuid>,
    pub door_to_doctor_mins: Option<i32>,
    pub door_to_disposition_mins: Option<i32>,
    pub vitals: Option<serde_json::Value>,
    pub notes: Option<String>,
    pub mass_casualty_event_id: Option<Uuid>,
    pub created_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErTriageAssessment {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Uuid,
    pub triage_level: TriageLevel,
    pub triage_system: String,
    pub score: Option<i32>,
    pub respiratory_rate: Option<i32>,
    pub pulse_rate: Option<i32>,
    pub blood_pressure_systolic: Option<i32>,
    pub blood_pressure_diastolic: Option<i32>,
    pub spo2: Option<i32>,
    pub gcs_score: Option<i32>,
    pub gcs_eye: Option<i32>,
    pub gcs_verbal: Option<i32>,
    pub gcs_motor: Option<i32>,
    pub pain_score: Option<i32>,
    pub chief_complaint: Option<String>,
    pub presenting_symptoms: Option<serde_json::Value>,
    pub allergies: Option<serde_json::Value>,
    pub is_pregnant: Option<bool>,
    pub disability_assessment: Option<String>,
    pub notes: Option<String>,
    pub assessed_by: Option<Uuid>,
    pub assessed_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErResuscitationLog {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Uuid,
    pub log_type: String,
    pub timestamp: DateTime<Utc>,
    pub medication_name: Option<String>,
    pub dose: Option<String>,
    pub route: Option<String>,
    pub fluid_name: Option<String>,
    pub fluid_volume_ml: Option<i32>,
    pub procedure_name: Option<String>,
    pub procedure_notes: Option<String>,
    pub vitals_snapshot: Option<serde_json::Value>,
    pub notes: Option<String>,
    pub recorded_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErCodeActivation {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Option<Uuid>,
    pub code_type: String,
    pub activated_at: DateTime<Utc>,
    pub deactivated_at: Option<DateTime<Utc>>,
    pub location: Option<String>,
    pub response_team: Option<serde_json::Value>,
    pub crash_cart_checklist: Option<serde_json::Value>,
    pub outcome: Option<String>,
    pub notes: Option<String>,
    pub activated_by: Option<Uuid>,
    pub deactivated_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MlcCase {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Option<Uuid>,
    pub patient_id: Uuid,
    pub mlc_number: String,
    pub status: MlcStatus,
    pub case_type: Option<String>,
    pub fir_number: Option<String>,
    pub police_station: Option<String>,
    pub brought_by: Option<String>,
    pub informant_name: Option<String>,
    pub informant_relation: Option<String>,
    pub informant_contact: Option<String>,
    pub history_of_incident: Option<String>,
    pub examination_findings: Option<String>,
    pub medical_opinion: Option<String>,
    pub is_pocso: bool,
    pub is_death_case: bool,
    pub cause_of_death: Option<String>,
    pub registered_by: Option<Uuid>,
    pub registered_at: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MlcDocument {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub mlc_case_id: Uuid,
    pub document_type: String,
    pub title: String,
    pub body_diagram: Option<serde_json::Value>,
    pub content: serde_json::Value,
    pub generated_by: Option<Uuid>,
    pub verified_by: Option<Uuid>,
    pub verified_at: Option<DateTime<Utc>>,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MlcPoliceIntimation {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub mlc_case_id: Uuid,
    pub intimation_number: String,
    pub police_station: String,
    pub officer_name: Option<String>,
    pub officer_designation: Option<String>,
    pub officer_contact: Option<String>,
    pub sent_at: DateTime<Utc>,
    pub sent_via: Option<String>,
    pub receipt_confirmed: bool,
    pub receipt_confirmed_at: Option<DateTime<Utc>>,
    pub receipt_number: Option<String>,
    pub notes: Option<String>,
    pub sent_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct MassCasualtyEvent {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub event_name: String,
    pub event_type: Option<String>,
    pub status: MassCasualtyStatus,
    pub activated_at: DateTime<Utc>,
    pub deactivated_at: Option<DateTime<Utc>>,
    pub location: Option<String>,
    pub estimated_casualties: Option<i32>,
    pub actual_casualties: Option<i32>,
    pub triage_summary: Option<serde_json::Value>,
    pub resources_deployed: Option<serde_json::Value>,
    pub notifications_sent: Option<serde_json::Value>,
    pub notes: Option<String>,
    pub activated_by: Option<Uuid>,
    pub deactivated_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Structured ER discharge summary (one per visit, draft → finalized).
/// Mirrors the IPD discharge summary; ER is acute and not encounter/
/// diagnosis-backed, so the clinical fields are free text.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErDischargeSummary {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Uuid,
    pub status: crate::ipd::DischargeSummaryStatus,
    pub final_diagnosis: Option<String>,
    pub condition_at_discharge: Option<String>,
    pub clinical_course: Option<String>,
    pub treatment_given: Option<String>,
    pub medications_on_discharge: Option<String>,
    pub follow_up_instructions: Option<String>,
    pub follow_up_date: Option<chrono::NaiveDate>,
    pub warning_signs: Option<String>,
    pub prepared_by: Option<Uuid>,
    pub verified_by: Option<Uuid>,
    pub finalized_at: Option<DateTime<Utc>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// One timestamped ER observation-chart entry — serial vitals + note
/// while a patient is held under observation before disposition.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErObservationNote {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub er_visit_id: Uuid,
    pub observed_at: DateTime<Utc>,
    pub pulse: Option<i32>,
    pub bp_systolic: Option<i32>,
    pub bp_diastolic: Option<i32>,
    pub resp_rate: Option<i32>,
    pub spo2: Option<i32>,
    pub temperature: Option<f64>,
    pub gcs: Option<i32>,
    pub pain_score: Option<i32>,
    pub note: Option<String>,
    pub recorded_by: Option<Uuid>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Configurable ER bay (master). Visits reference a bay by `code`; the
/// occupancy board is computed from active visits.
#[derive(Debug, Clone, Serialize, Deserialize, FromRow)]
pub struct ErBay {
    pub id: Uuid,
    pub tenant_id: Uuid,
    pub code: String,
    pub name: String,
    pub bay_type: Option<String>,
    pub is_active: bool,
    pub sort_order: i32,
    pub notes: Option<String>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

File: medbrains/packages/types/src/emergency.ts
// Emergency (ER) types — split from index.ts, barrel-re-exported.

// ── Emergency ──────────────────────────────────────────

export type TriageLevel =
  | "immediate"
  | "emergent"
  | "urgent"
  | "less_urgent"
  | "non_urgent"
  | "expectant"
  | "unassigned";
export type ErVisitStatus =
  | "registered"
  | "triaged"
  | "in_treatment"
  | "observation"
  | "admitted"
  | "discharged"
  | "transferred"
  | "lama"
  | "deceased";
export type MlcStatus =
  | "registered"
  | "under_investigation"
  | "opinion_given"
  | "court_pending"
  | "closed";
export type MassCasualtyStatus = "activated" | "ongoing" | "scaling_down" | "deactivated";

export interface MailDnsRecord {
  record_type: string;
  host: string;
  value: string;
}

export interface ProvisionDomainResponse {
  domain: string;
  dns_records: MailDnsRecord[];
}

export interface PatientRadiologyReport {
  report_id: string;
  order_id: string;
  modality: string | null;
  ordered_at: string;
  status: string;
  findings: string;
  impression: string | null;
  is_critical: boolean;
  verified_at: string | null;
  alert_id: string | null;
  alert_acknowledged_at: string | null;
}

export interface PatientTimelineEvent {
  occurred_at: string;
  category: "opd" | "ipd" | "emergency" | "lab" | "radiology" | "pharmacy" | "diagnosis";
  title: string;
  subtitle: string | null;
  ref_id: string;
}

export interface ErBay {
  id: string;
  tenant_id: string;
  code: string;
  name: string;
  bay_type: string | null;
  is_active: boolean;
  sort_order: number;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface ErBayRequest {
  code: string;
  name: string;
  bay_type?: string;
  is_active?: boolean;
  sort_order?: number;
  notes?: string;
}

export interface ErObservationNote {
  id: string;
  tenant_id: string;
  er_visit_id: string;
  observed_at: string;
  pulse: number | null;
  bp_systolic: number | null;
  bp_diastolic: number | null;
  resp_rate: number | null;
  spo2: number | null;
  temperature: number | null;
  gcs: number | null;
  pain_score: number | null;
  note: string | null;
  recorded_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface CreateObservationNoteRequest {
  pulse?: number;
  bp_systolic?: number;
  bp_diastolic?: number;
  resp_rate?: number;
  spo2?: number;
  temperature?: number;
  gcs?: number;
  pain_score?: number;
  note?: string;
}

export interface ErVisit {
  id: string;
  tenant_id: string;
  patient_id: string;
  encounter_id: string | null;
  visit_number: string;
  status: ErVisitStatus;
  arrival_mode: string | null;
  arrival_time: string;
  chief_complaint: string | null;
  is_mlc: boolean;
  is_brought_dead: boolean;
  triage_level: TriageLevel | null;
  attending_doctor_id: string | null;
  bay_number: string | null;
  disposition: string | null;
  disposition_time: string | null;
  disposition_notes: string | null;
  admitted_to: string | null;
  admission_id: string | null;
  door_to_doctor_mins: number | null;
  door_to_disposition_mins: number | null;
  vitals: Record<string, unknown> | null;
  notes: string | null;
  mass_casualty_event_id: string | null;
  created_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface ErTriageAssessment {
  id: string;
  tenant_id: string;
  er_visit_id: string;
  triage_level: TriageLevel;
  triage_system: string;
  score: number | null;
  respiratory_rate: number | null;
  pulse_rate: number | null;
  blood_pressure_systolic: number | null;
  blood_pressure_diastolic: number | null;
  spo2: number | null;
  gcs_score: number | null;
  gcs_eye: number | null;
  gcs_verbal: number | null;
  gcs_motor: number | null;
  pain_score: number | null;
  chief_complaint: string | null;
  presenting_symptoms: Record<string, unknown> | null;
  allergies: Record<string, unknown> | null;
  is_pregnant: boolean | null;
  disability_assessment: string | null;
  notes: string | null;
  assessed_by: string | null;
  assessed_at: string;
  created_at: string;
  updated_at: string;
}

export interface ErResuscitationLog {
  id: string;
  tenant_id: string;
  er_visit_id: string;
  log_type: string;
  timestamp: string;
  medication_name: string | null;
  dose: string | null;
  route: string | null;
  fluid_name: string | null;
  fluid_volume_ml: number | null;
  procedure_name: string | null;
  procedure_notes: string | null;
  vitals_snapshot: Record<string, unknown> | null;
  notes: string | null;
  recorded_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface ErCodeActivation {
  id: string;
  tenant_id: string;
  er_visit_id: string | null;
  code_type: string;
  activated_at: string;
  deactivated_at: string | null;
  location: string | null;
  response_team: Record<string, unknown> | null;
  crash_cart_checklist: Record<string, unknown> | null;
  outcome: string | null;
  notes: string | null;
  activated_by: string | null;
  deactivated_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface MlcCase {
  id: string;
  tenant_id: string;
  er_visit_id: string | null;
  patient_id: string;
  mlc_number: string;
  status: MlcStatus;
  case_type: string | null;
  fir_number: string | null;
  police_station: string | null;
  brought_by: string | null;
  informant_name: string | null;
  informant_relation: string | null;
  informant_contact: string | null;
  history_of_incident: string | null;
  examination_findings: string | null;
  medical_opinion: string | null;
  is_pocso: boolean;
  is_death_case: boolean;
  cause_of_death: string | null;
  registered_by: string | null;
  registered_at: string;
  created_at: string;
  updated_at: string;
}

export interface MlcDocument {
  id: string;
  tenant_id: string;
  mlc_case_id: string;
  document_type: string;
  title: string;
  body_diagram: Record<string, unknown> | null;
  content: Record<string, unknown>;
  generated_by: string | null;
  verified_by: string | null;
  verified_at: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
}

export interface MlcPoliceIntimation {
  id: string;
  tenant_id: string;
  mlc_case_id: string;
  intimation_number: string;
  police_station: string;
  officer_name: string | null;
  officer_designation: string | null;
  officer_contact: string | null;
  sent_at: string;
  sent_via: string | null;
  receipt_confirmed: boolean;
  receipt_confirmed_at: string | null;
  receipt_number: string | null;
  notes: string | null;
  sent_by: string | null;
  created_at: string;
  updated_at: string;
}

export interface MassCasualtyEvent {
  id: string;
  tenant_id: string;
  event_name: string;
  event_type: string | null;
  status: MassCasualtyStatus;
  activated_at: string;
  deactivated_at: string | null;
  location: string | null;
  estimated_casualties: number | null;
  actual_casualties: number | null;
  triage_summary: Record<string, unknown> | null;
  resources_deployed: Record<string, unknown> | null;
  notifications_sent: Record<string, unknown> | null;
  notes: string | null;
  activated_by: string | null;
  deactivated_by: string | null;
  created_at: string;
  updated_at: string;
}
